feat: `kube-aws up --s3-uri s3://<bucket>/<directory> to automatically avoid the 51200 bytes limitation errors of CloudFormation

mumoshu opened the pull request:
If you are hit by the cloudformation limit, `kube-aws up` now fail with a specific error explaining the limit and how to work-around it:

```
$ kube-aws up
Creating AWS resources. This should take around 5 minutes.
Error: Error creating cluster: stack-template.json size(=51673) exceeds the 51200 bytes limit of cloudformation. `--s3-uri s3://<bucket>/path/to/dir` must be specified to upload it to S3 beforehand
```

As the error message says, if you provide the `--s3-uri` option composed of a s3 bucket's name and a path to directory, `kube-aws` uploads your template to s3 if necessary:

```
$ kube-aws up --s3-uri s3://mybucket/mydir
```

resolves #38 (for now)

mumoshu commented:
TODO
- [ ] `kube-aws update` support (it doesn't respect `--s3-uri` now)
- [ ] `kube-aws validate` fix (it doesn't accept `--s3-uri` so it always complain when you're exceeding the 51200 bytes limit)
- [ ] unit tests

Does anyone need those TODOs to be addressed before merging this, or you'd like to merge this asap, anything to be added to TODOs, etc?

Any feedback is welcome 🙏

pieterlange commented:
Cool! Thanks for the fast turnaround on this @mumoshu!

I think we shouldn't permit people to paint themselves into a corner by not supporting `update`, let's take some time to fully implement this :+1:

mumoshu commented:
@pieterlange Thanks for the first feedback 👍 

According to your feedback, I have just fixed  `kube-aws update` to support `--s3-uri`.

To everyone: I'm going to run e2e covering `kube-aws update` eventually but in the meantime, I'd appreciate it if you could test `kube-aws update --s3-uri` in your environments and give feedbacks!

camilb commented:
@mumoshu Now works fine with `kube-aws up --s3-uri s3://<bucket>/<directory>` and
`kube-aws update --s3-uri s3://<bucket>/<directory>`

    Update stack: {
      StackId: "arn:aws:cloudformation:us-east-1:xxx:stack/kube/xxxx"
    }
    Success! Your AWS resources are being updated:
    %!(EXTRA string=Cluster Name:		kube
    Controller DNS Name:	xxx.us-east-1.elb.amazonaws.com
    )

mumoshu commented:
@camilb Thanks!

For anyone interested, I've:

* rebased this to the latest master,
* added `kube-aws validate --s3-uri` to validate your stack template when its larger than the limit.

I've also tested `validate`, `up`, and then `update` with my e2e script and verified it to work.

mumoshu commented:
Some tests added.
Note that now, it also accept URIs without directories i.e. `--s3-uri s3://mybucket` and `s3-uri s3://mybucket/mydir` are both valid.

mumoshu commented:
Addressed all the TODOs and E2E, unit tess are passing. Merging.

mumoshu commented:
I'm just merging this to get the ball rolling.
Feel free to reopen #38 and/or submit another issue(s) for feedbacks!
